router/dataplane_spec.gobra

```diff
@@ -343,3 +349,262 @@ func closureSpec2(i uint16, c BatchConn)
 requires true
 func closureSpec3(c BatchConn)
 /** End of closure specs for the Run method **/
+
+
+
+/** Start of io-spec helper functions **/
+
+
+ghost 
+requires seg1Len > 0
+requires seg2Len >= 0
+requires seg3Len >= 0
+requires currHF < seg1Len + seg2Len + seg3Len
+decreases
+pure func computeSeqlen(currHF int, seg1Len int, seg2Len int, seg3Len int) int{
+	return seg1Len > currHF ? seg1Len : ((seg1Len + seg2Len) > currHF ?  seg2Len: seg3Len)
+}
+
+ghost 
+decreases
+pure func infoFieldOffsetSeg(currINF int) int{
+	return 4 + 8 * currINF
+}
+
+ghost 
+decreases 
+pure func numInfoFields(seg1Len int, seg2Len int, seg3Len int) int {
+	return seg3Len > 0 ? 3 : (seg2Len > 0 ? 2 : 1)
+}
+
+ghost 
+requires seg1Len > 0
+requires seg2Len >= 0
+requires seg3Len >= 0
+requires currHF >= 0
+ensures res <= currHF
+decreases
+pure func computeprevSegLen(currHF int, seg1Len int, seg2Len int, seg3Len int) (res int){
+	return seg1Len > currHF ? 0 : ((seg1Len + seg2Len) > currHF ? seg1Len : seg1Len + seg2Len)
+}
+
+//expects the raw bytes of whole path
+ghost
+requires start + 12 <= len(raw)
+requires 0 <= start
+requires acc(&raw[start+2], _) && acc(&raw[start+3], _) && acc(&raw[start+4], _) && acc(&raw[start+5], _)
+decreases
+pure func getHopField(raw []byte, start int) io.IO_HF {
+	return let inif2 := binary.BigEndian.Uint16(raw[start+2:start+4]) == 0 ? none[io.IO_ifs] : some(io.IO_ifs(binary.BigEndian.Uint16(raw[start+2:start+4]))) in
+			let egif2 := binary.BigEndian.Uint16(raw[start+4:start+6])== 0 ? none[io.IO_ifs] : some(io.IO_ifs(binary.BigEndian.Uint16(raw[start+4:start+6]))) in
```

Comment:
these lines are quite long and hard to read. please, try to extract repeated subexpressions to new let bindings and make sure that all let bindings are aligned

Reply:
Done.